packages/@react-aria/selection/src/useSelectableCollection.ts

```diff
@@ -401,21 +405,52 @@ export function useSelectableCollection(options: AriaSelectableCollectionOptions
 
     // If the user is typing forwards, autofocus the first option in the list.
     if (detail?.focusStrategy === 'first') {
-      let keyToFocus = delegate.getFirstKey?.() ?? null;
-      // If no focusable items exist in the list, make sure to clear any activedescendant that may still exist
-      if (keyToFocus == null) {
-        ref.current?.dispatchEvent(
-          new CustomEvent(UPDATE_ACTIVEDESCENDANT, {
-            cancelable: true,
-            bubbles: true
-          })
-        );
-      }
+      shouldVirtualFocusFirst.current = true;
+    }
+  });
 
+  let updateActiveDescendant = useEffectEvent(() => {
+    let keyToFocus = delegate.getFirstKey?.() ?? null;
+
+    // If no focusable items exist in the list, make sure to clear any activedescendant that may still exist
+    if (keyToFocus == null) {
+      ref.current?.dispatchEvent(
+        new CustomEvent(UPDATE_ACTIVEDESCENDANT, {
+          cancelable: true,
+          bubbles: true
+        })
+      );
+    } else {
       manager.setFocusedKey(keyToFocus);
+      // Only set shouldVirtualFocusFirst to false if we've successfully set the first key as the focused key
+      // If there wasn't a key to focus, we might be in a temporary loading state so we'll want to still focus the first key
+      // after the collection updates after load
+      shouldVirtualFocusFirst.current = false;
+    }
+  });
+
+  let lastCollection = useRef(manager.collection);
+  useEffect(() => {
```

Comment:
You might be able to use `useUpdateEffect` here as well as it will do the comparison of collection for you (so you don't need to store it in a ref yourself).